src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataDictionaryFormatter.cs

```diff
@@ -0,0 +1,363 @@
+    {
+        public static readonly MetadataDictionaryFormatter Instance = new();
+
+        private MetadataDictionaryFormatter()
+        {
+        }
+
+        /// <inheritdoc/>
+        public void Serialize(ref MessagePackWriter writer, IReadOnlyDictionary<string, object?> value, MessagePackSerializerOptions options)
+        {
+            writer.Write(value.Count);
+
+            // Special case certain values to avoid defeating lazy load later. Check out the
+            // ReadMetadata below, how it wraps the return value.
+            IReadOnlyDictionary<string, object?> serializedMetadata;
+
+            // Unwrap the metadata if its an instance of LazyMetaDataWrapper, the wrapper may end up
+            // implicitly resolving TypeRefs to Types which is undesirable.
+            value = LazyMetadataWrapper.TryUnwrap(value);
+            serializedMetadata = new LazyMetadataWrapper(value.ToImmutableDictionary(), LazyMetadataWrapper.Direction.ToSubstitutedValue, options.CompositionResolver());
+
+            foreach (KeyValuePair<string, object?> item in serializedMetadata)
+            {
+                writer.Write(item.Key);
+
+                SerializeObject(ref writer, item.Value);
+            }
+
+            void SerializeObject(ref MessagePackWriter messagePackWriter, object? value)
+            {
+                if (value is null)
+                {
+                    messagePackWriter.Write((byte)MetadataDictionaryFormatter.ObjectType.Null);
+                    return;
+                }
+
+                switch (value.GetType())
+                {
+                    case Type objectType when objectType.IsArray:
+
+                        var array = (Array)value;
+                        messagePackWriter.Write((byte)MetadataDictionaryFormatter.ObjectType.Array);
+
+                        TypeRef? elementTypeRef = TypeRef.Get(objectType.GetElementType(), options.CompositionResolver());
+                        options.Resolver.GetFormatterWithVerify<TypeRef?>().Serialize(ref messagePackWriter, elementTypeRef, options);
+                        messagePackWriter.Write(array.Length);
+                        foreach (object? item in array)
+                        {
+                            SerializeObject(ref messagePackWriter, item);
+                        }
+
+                        break;
+
+                    case Type objectType when objectType == typeof(bool):
+                        ObjectType objectValueType = (bool)value ? ObjectType.BoolTrue : ObjectType.BoolFalse;
+                        messagePackWriter.Write((byte)objectValueType);
+                        break;
+
+                    case Type objectType when objectType == typeof(string):
+                        messagePackWriter.Write((byte)ObjectType.String);
+                        messagePackWriter.Write((string)value);
+                        break;
```

Comment:
You're writing twice for a single value, which breaks msgpack parsers. I suggest we fix this with a two pronged approach:
1. Don't use the `ObjectType` enum for any type for which msgpack has a matching primitive, because such primitive types can be detected while reading by virtue of the `MessagePackReader`, making the enum redundant for all those cases. Take care though when you write though that you use the explicit type named methods (e.g. `WriteUInt32`) instead of just `Write(uint)`, or else msgpack's type annotation will be 'lossy', which we can't afford. In the reader, you can build a switch statement around `reader.NextMessagePackType`.
2. For the types for which msgpack doesn't have a matching primitive, you'll need to call `WriteArrayHeader(2)` followed by the `ObjectType` enum value, and finally the metadatum value itself.